[PeterHW963] iP

PeterHW963 opened the pull request:
### Nether

> “Your mind is for having ideas, not holding them.” – David Allen

Nether frees your mind of having to remember things you need to do. It's
- text-based
- easy to learn
- ~~FAST~~ _SUPER FAST_ to use

All you need to do is
1. download it from [here](https://nus-cs2103-ay2425s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you 😉 

And it is **FREE**!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

---

If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

PeterHW963 committed:
Change Duke.java to Nether.java

Duke.java visually prints out "Duke"

The assignment given is to rename the chatbot, and create a greet and
exit message.

Let's rename the file to Nether.java, its new name and create the
greet and exit messages, along with a new visual print out of its name.

"println" statements are the only lines edited as that is the only
requirement for this increment.

PeterHW963 committed:
Add Echo feature until user inputs "bye"

Chatbot produces visual art of its name, greetings, and goodbye message
when run.

The assignment given is to have chatbot produce echo whenever the
user inputs anything other than "bye".

Let's create a Scanner object to receive user input and create an
infinite loop of echoing the user input and listening for more input. We
used a check inside the infinite loop for a "bye" input to break out of
the loop and produce the goodbye message.

PeterHW963 committed:
Add "Add" and "List" features and refactored code

Chatbot echoes user input unless the user inputs "bye"

The assignment given is to have the chatbot add whatever the user inputs
into a list of tasks and to add another check for the input "list", to
which the chatbot should return a numbered list of all the tasks the
user has input so far.

Let's create an array of size 100 to store the tasks input by user along
with a counter to keep track of the current index in the array. When
the user inputs "list", a for loop is used to iterate through the array
and print out the entries.

Additionally, an abstraction for printing the horizontal line output is
made and the constants MAX_TASKS, EXIT_COMMAND, and LIST_COMMAND were
created to improve reusability and readability.

PeterHW963 committed:
Add Mark/Unmark as Done and refactor code

Chatbot add whatever the user inputs into a list of tasks and prints a
numbered list of all the tasks the user has input so far whenever the
user inputs "list".

The assignment given is to have the chatbot have the ability to
mark/unmark tasks as done/not done when the user inputs "mark +
(integer)" or "unmark + (integer)".

Let's create a new class called Task to represent tasks. Task has a
description, index, and an isDone status. Task class also has the
necessary getters and setters. Next, let's implement an extra check for
the user input (i.e. if the user input starts with "mark" or "unmark")
that modifies the task object to be done or not done respectively.

Additionally, an abstraction for displaying the task list is
made and the constants MARK_AS_DONE and MARK_AS_NOT_DONE were
created to improve reusability and readability.

A new Task class was created to encapsulate the data and behavior of
task objects, improve reusability, maintainability, and readability.

PeterHW963 committed:
Add ToDos, Events, Deadlines

Chatbot adds whatever the user inputs into a list of tasks and prints a
numbered list of all the tasks the user has input so far whenever the
user inputs "list" along with its done/not done status.

The assignment given is to have the chatbot have the ability to
recognize what type of task is added, whether it is ToDos, Deadline, or
Event type tasks and have the task type displayed when listing the
tasks.

Let's create new subclasses for the Task class to represent the 3 types
of tasks. DeadlineTask has a by attribute and EventTask has from and to
attributes. Each of these Task subclasses have their own constructor and
toString methods. Next, let's implement a method that parses the user
input into the correct number of parts to instantiate the respective
tasks with.

Inheritance was used to support multiple task types since the different
tasks have some similarity between them.

PeterHW963 committed:
Add Error handling and text UI testing

Chatbot adds one of the 3 types of tasks the user inputs into a
numbered list. The tasks can be marked as done or not done, and user can
input "list" to view all the tasks written so far.

The assignment given is to have the chatbot handle errors, especially
when the user enters an unexpected input that the chatbot is not
programmed to handle.

Let's create a new exception called NetherException that can produce the
corresponding error message. Additionally, an automated testing has been
implemented.

PeterHW963 committed:
Add delete task

Chatbot adds one of the 3 types of tasks the user inputs into a
numbered list. The tasks can be marked as done or not done, and user can
input "list" to view all the tasks written so far. Chatbot can also
handle various inputs from the user.

The assignment given is to have the chatbot able to receive delete
commands to delete a task from the list.

Let's add an extra check to the chatbot for a "delete" command to then
remove the task from the list

PeterHW963 committed:
Add Save data

Chatbot is initialized with a blank task list every time.

The assignment given is to have the chatbot able to save task list data
and load task list from existing data (if any) upon initialization.

Let's create a new Storage class that handles the saving and loading of
tasks from the data file. Also let's implement a method toSaveFormat
for each task type that processes the existing task list data into a
form that is easily processed later on for loading.

PeterHW963 committed:
Attempt to rectify branching mistake

PeterHW963 committed:
Merge branch 'branch-level-7'

PeterHW963 committed:
Add Dates and Times

Chatbot can take any date and/or time format from user inputs.

The assignment given is to have the chatbot able to validate input
date/time and save task date/time using LocalDateTime. Additionally,
the toString() method should also present the date/time of tasks in a
more user-friendly way.

Let's make use of the LocalDateTime and DateTimeFormatter to parse user
date/time input, and later on, revert it into String when it needs to be
printed for users to see.

PeterHW963 committed:
Merge branch 'branch-level-8'

PeterHW963 committed:
Refactor code using more OOP

PeterHW963 committed:
Merge branch 'branch-moreOOP'

PeterHW963 committed:
Organize into packages

PeterHW963 committed:
Merge branch 'branch-packages'

PeterHW963 committed:
Merge branch 'add-gradle-support'

PeterHW963 committed:
Change gradle build configuration

Build configuration for gradle is default (i.e. mainClass was the Duke
class)

Let's change the mainClass to be the nether.Nether class (specific to
my own iP)

PeterHW963 committed:
Merge branch 'branch-gradle'

PeterHW963 committed:
Add JUnit Tests

PeterHW963 committed:
Merge branch 'branch-JUnit'

PeterHW963 committed:
Create a JAR File

PeterHW963 committed:
Merge branch 'branch-Jar'

PeterHW963 committed:
Add JavaDoc style documentation

PeterHW963 committed:
Follow the Coding Standard

PeterHW963 committed:
Add find function

PeterHW963 committed:
Merge branch 'branch-A-JavaDoc'

PeterHW963 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/nether/command/MarkNotDoneCommand.java

PeterHW963 committed:
Merge branch 'branch-Level-9'

PeterHW963 committed:
Use CheckStyle

Checkstyle configuration setup and is now running

PeterHW963 committed:
Merge branch 'branch-checkstyle'

PeterHW963 committed:
Prepare for GUI implementation

PeterHW963 committed:
Add GUI

Replaced the Text UI with the GUI.
Still have some more GUI fixes to do afterwards.

PeterHW963 committed:
Merge branch 'branch-level-10'

PeterHW963 committed:
Change style of Window and misc changes

yyihaoc commented on src/main/java/nether/command/Command.java:
```diff
@@ -0,0 +1,40 @@
+ *
+ */
+public abstract class Command {
+    /**
+     * Executes the command with the provided task list, user interface, and storage.
+     *
+     * This method must be implemented by subclasses to define the specific behavior of the command.
+     *
+     * @param tasks The {@code TaskList} to be modified or queried by the command.
+     * @param ui The {@code Ui} instance used to interact with the user.
+     * @param storage The {@code Storage} instance used to persist changes.
+     * @throws NetherException If an error occurs during command execution.
+     */
```
I like how comprehensive these method comments are!

yyihaoc commented on src/main/java/nether/command/MarkNotDoneCommand.java:
```diff
@@ -0,0 +1,48 @@
+     * @return The message "Understood, I've marked this task as not done:".
+     */
+    @Override
+    public String getMarkMessage() {
```
I like the naming!

yyihaoc commented on src/main/java/nether/command/FindCommand.java:
```diff
@@ -0,0 +1,24 @@
+ * The search is case-insensitive and returns a list of all tasks that contain the input string.
+ */
+public class FindCommand extends Command {
+    private final String searchString;
```
Good and concise naming!

yyihaoc commented on src/main/java/nether/command/ListCommand.java:
```diff
@@ -0,0 +1,30 @@
+     */
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) {
+        return tasks.printList();
```
Not sure if this actually prints to screen the tasks. I might have overlooked.

yyihaoc commented:
LGTM!

PeterHW963 committed:
Ensure JUnit Tests are still passed

A lot of checkstyle tests were failing.

Followed suggestions to correct all of them.

PeterHW963 committed:
Create guard case for list command

There is no special response message when the 'list' command is
input with an empty task list.

Creating a special response message makes the chatbot response more
realistic.

Let's create a guard case using an if statement to return a special
response string when 'list' is input with an empty task list.

PeterHW963 committed:
Add assertions

Abnormal user input such as inputting an empty string has not been
handled.

Assertion for this particular input is implemented.

Let's use assert !Object.equals("", userInput) to then give the
appropriate error, though it will not appear in the GUI.

PeterHW963 committed:
Improve Code Quality

Code has lapses in following the Code Quality given in this week's
materials.

The principles below has been followed now:
- avoid long methods
- avoid deep nesting
- avoid complicated expressions
- avoid magic numbers
- make the code obvious
- structure code logically
- do not 'trip up' reader
- practice Keep it simple, stupid
- Single Level of Abstraction Principle
- Make the happy path prominent
- don't recycle variables or parameters
- use default branchs
- avoid empty catch blocks
- delete dead code
- Minimize scope of variables
- Minimize code duplication

PeterHW963 committed:
Use streams

PeterHW963 committed:
Merge pull request #2 from PeterHW963/branch-A-CodeQuality

Improve Code Quality

PeterHW963 committed:
Merge branch 'master' into branch-A-Assertions

PeterHW963 committed:
Merge pull request #1 from PeterHW963/branch-A-Assertions

Add assertions

PeterHW963 committed:
Solve issues with data file again

PeterHW963 committed:
Merge branch 'master' into branch-A-Streams

PeterHW963 committed:
Merge pull request #3 from PeterHW963/branch-A-Streams

Use streams

PeterHW963 committed:
Merge branch 'master' of https://github.com/PeterHW963/ip

# Conflicts:
#	data/nether.txt

PeterHW963 committed:
Set up CI

PeterHW963 committed:
Merge branch 'Branch-A-CI'

PeterHW963 committed:
Fix .yaml file

PeterHW963 committed:
Copy FXML from guide

PeterHW963 committed:
Finish implementing CSS for GUI

PeterHW963 committed:
Add C-Tagging extension

PeterHW963 committed:
Implement Better GUI

PeterHW963 committed:
Add a unique personality

PeterHW963 committed:
Fix style for CI

PeterHW963 committed:
Fix for CI (pt2)

PeterHW963 committed:
Fix for CI and upload Ui.png

PeterHW963 committed:
Fix CI (pt4)

PeterHW963 committed:
Add tag finding and start README

PeterHW963 committed:
Update README.md

PeterHW963 committed:
Ready for release

Updated README and added exit functionality.

PeterHW963 committed:
WALLAH LAST FIX FOR CI

PeterHW963 committed:
Remove old README and add acknowledgements to current README
